parser/parser_utils.go

```diff
@@ -0,0 +1,167 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/datastax/go-cassandra-native-protocol/datatype"
+	"github.com/datastax/go-cassandra-native-protocol/message"
+)
+
+const (
+	CountValueName = "count(*)"
+)
+
+var systemTables = []string{"local", "peers", "peers_v2", "schema_keyspaces", "schema_columnfamilies", "schema_columns", "schema_usertypes"}
+
+type ValueLookupFunc func(name string) (value message.Column, err error)
+
+func FilterValues(stmt *SelectStatement, columns []*message.ColumnMetadata, valueFunc ValueLookupFunc) (filtered []message.Column, err error) {
+	if _, ok := stmt.Selectors[0].(*StarSelector); ok {
+		for _, column := range columns {
+			var val message.Column
+			val, err = valueFunc(column.Name)
+			if err != nil {
+				return nil, err
+			}
+			filtered = append(filtered, val)
+		}
+	} else {
+		for _, selector := range stmt.Selectors {
+			var val message.Column
+			val, err = valueFromSelector(selector, valueFunc)
+			if err != nil {
+				return nil, err
+			}
+			filtered = append(filtered, val)
+		}
+	}
+	return filtered, nil
+}
+
+func valueFromSelector(selector Selector, valueFunc ValueLookupFunc) (val message.Column, err error) {
+	switch s := selector.(type) {
+	case *CountStarSelector:
+		return valueFunc(CountValueName)
+	case *IDSelector:
+		return valueFunc(s.Name)
+	case *AliasSelector:
+		return valueFromSelector(s.Selector, valueFunc)
+	default:
+		return nil, errors.New("unhandled selector type")
+	}
+}
+
+func FilterColumns(stmt *SelectStatement, columns []*message.ColumnMetadata) (filtered []*message.ColumnMetadata, err error) {
+	if _, ok := stmt.Selectors[0].(*StarSelector); ok {
+		filtered = columns
+	} else {
+		for _, selector := range stmt.Selectors {
+			var column *message.ColumnMetadata
+			column, err = columnFromSelector(selector, columns, stmt.Table)
+			if err != nil {
+				return nil, err
+			}
+			filtered = append(filtered, column)
+		}
+	}
+	return filtered, nil
+}
+
+func isCountSelector(selector Selector) bool {
+	_, ok := selector.(*CountStarSelector)
+	return ok
+}
+
+func IsCountStarQuery(stmt *SelectStatement) bool {
+	if len(stmt.Selectors) == 1 {
+		if isCountSelector(stmt.Selectors[0]) {
+			return true
+		} else if alias, ok := stmt.Selectors[0].(*AliasSelector); ok {
+			return isCountSelector(alias.Selector)
+		}
+	}
+	return false
+}
+
+func columnFromSelector(selector Selector, columns []*message.ColumnMetadata, table string) (column *message.ColumnMetadata, err error) {
+	switch s := selector.(type) {
+	case *CountStarSelector:
+		return &message.ColumnMetadata{
+			Keyspace: "system",
```

Comment:
I'm probably missing something but why is this always `system` keyspace?

Reply:
Currently, all handled queries are in the "system" keyspace. I agree this doesn't really belong here. I've added `Keyspace` to the `SelectStatement` type so this decision isn't made here anymore.